common/network-types/src/session/processing/mod.rs

```diff
@@ -0,0 +1,254 @@
+    impl futures::Stream<Item = Result<Frame, SessionError>>,
+) {
+    use futures::StreamExt;
+
+    let (sink, rs_stream) = reassembler.split();
+    let (seq_sink, stream) = sequencer.split();
+
+    let id = id.to_string();
+    hopr_async_runtime::prelude::spawn(
+        async {
+            match rs_stream
+                .filter_map(|maybe_frame| async {
+                    // Frames that fail to reassemble will eventually be
+                    // discarded in the sequencer as missing,
+                    // so we're safe to filter them out here and only log them.
+                    maybe_frame
+                        .inspect_err(|error| tracing::error!(%error, "failed to reassemble frame"))
+                        .ok()
+                        .map(|f| Ok(OrderedFrame(f)))
+                })
+                .forward(seq_sink)
+                .await
+            {
+                Ok(_) => tracing::debug!("frame reconstructor finished"),
+                Err(error) => tracing::error!(%error, "frame reconstructor finished with error"),
+            }
+        }
+        .instrument(tracing::debug_span!("FrameReconstructor", session_id = %id)),
+    );
+    (sink, stream.map(|f| f.map(|of| of.0)))
+}
+
+pub fn frame_reconstructor(
+    id: &str,
+    frame_timeout: std::time::Duration,
+    capacity: usize,
+) -> (
+    impl futures::Sink<Segment, Error = SessionError>,
+    impl futures::Stream<Item = Result<Frame, SessionError>>,
+) {
+    build_reconstructor(
+        id,
+        Reassembler::<FrameHashMap>::new(frame_timeout, capacity),
+        Sequencer::new(sequencer::SequencerConfig {
+            timeout: frame_timeout,
+            capacity,
+            ..Default::default()
+        }),
+    )
+}
+
+pub fn frame_reconstructor_with_inspector(
+    id: &str,
+    frame_timeout: std::time::Duration,
+    capacity: usize,
+) -> (
+    impl futures::Sink<Segment, Error = SessionError>,
+    impl futures::Stream<Item = Result<Frame, SessionError>>,
+    FrameInspector,
+) {
+    let reassembler = Reassembler::<FrameDashMap>::new(frame_timeout, capacity);
+    let inspector = reassembler.new_inspector();
+
+    let (sink, stream) = build_reconstructor(
+        id,
+        reassembler,
+        Sequencer::new(sequencer::SequencerConfig {
+            timeout: frame_timeout,
+            capacity,
+            ..Default::default()
+        }),
+    );
+
+    (sink, stream, inspector)
+}
+
+/// Helper function to segment `data` into segments of a given ` max_segment_size ` length.
+/// All segments are tagged with the same `frame_id`.
+#[cfg(test)]
+pub fn segment<T: AsRef<[u8]>>(
+    data: T,
+    max_segment_size: usize,
+    frame_id: u32,
+) -> crate::session::errors::Result<Vec<Segment>> {
+    use crate::session::frames::SeqNum;
+
+    if frame_id == 0 {
+        return Err(SessionError::InvalidFrameId);
+    }
+
+    if max_segment_size == 0 {
+        return Err(SessionError::InvalidSegmentSize);
+    }
```

Comment:
So why not make this a `NonZero`?

Reply:
I will check, but might not be most ergonomical.